### ISSUE TITLE
Error when trying to upload a file with weird non printable characters present

### ISSUE THREAD

**Issue by jose-sanchezm** [abraunegg/onedrive#35]
Hello, I'm trying to synchronize my documents from several projects to my OneDrive Business account. I've been able to upload most of them by using the command `onedrive --synchronize --upload-only --verbose` but at some point, it fails with the following message:

```
The file has not changed
Processing OneDrive_lastlogon-MC0XR79C.txt
The file has not changed
Processing seguridad.zip
The file has not changed
Processing OneDrive_lastlogon-MC0XR79C-2.txt
The file has not changed
Uploading new items of .
std.json.JSONException@std/json.d(1392): Unexpected character '<'. (Line 1:1)
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN""http://www.w3.org/TR/html4/strict.dtd">
<HTML><HEAD><TITLE>Bad Request</TITLE>
<META HTTP-EQUIV="Content-Type" Content="text/html; charset=us-ascii"></HEAD>
<BODY><h2>Bad Request - Invalid URL</h2>
<hr><p>HTTP Error 400. The request URL is invalid.</p>
</BODY></HTML>

----------------
??:? pure @safe void std.json.parseJSON!(char[]).parseJSON(char[], int, std.json.JSONOptions).error(immutable(char)[]) [0xb36e9530]
??:? pure @safe void std.json.parseJSON!(char[]).parseJSON(char[], int, std.json.JSONOptions).parseValue(ref std.json.JSONValue) [0xb36ea10b]
??:? pure @safe std.json.JSONValue std.json.parseJSON!(char[]).parseJSON(char[], int, std.json.JSONOptions) [0xb36e94cd]
??:? std.json.JSONValue onedrive.OneDriveApi.perform() [0xb36e8d2a]
??:? std.json.JSONValue onedrive.OneDriveApi.get(const(char)[], bool) [0xb36e85ae]
??:? std.json.JSONValue onedrive.OneDriveApi.getPathDetails(const(immutable(char)[])) [0xb36e7af3]
??:? void sync.SyncEngine.uploadNewFile(immutable(char)[]) [0xb36f581f]
??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0xb36f4f53]
??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0xb36f4e4a]
??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0xb36f4e4a]
??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0xb36f4e4a]
??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0xb36f4e4a]
??:? void sync.SyncEngine.scanForDifferences(immutable(char)[]) [0xb36f4019]
??:? void main.performSync(sync.SyncEngine, immutable(char)[], bool, bool, bool) [0xb36da557]
??:? _Dmain [0xb36d9ef7]

```

How can I know which file/directory is making it fail? Is there any way that I can debug it?

**Comment by abraunegg** [comment]
When you say big number of files - how many are we talking about here? This being said:

`std.json.JSONException@std/json.d(1392): Unexpected character '<'. (Line 1:1)`

this is the crux of the issue. What is happening is that OneDrive (graph or direct) is giving a garbled response. I have typically seen this mainly when just starting the application, but not during the actual upload - so that is 'new'.

Generally when you next run the application there is no issue - can you confirm that?

This being said - as this is a Business account #23 & #2 applies as well which would be impacting the uploads.

**Comment by jose-sanchezm** [comment]
According to Nautilus there are "665 items, totalling 1,0 GB" files in my OneDrive folder that I'm trying to upload and it fails consistently evey time that I run the `onedrive --synchronize --upload-only --verbose` command. It shows a lot of 
```
Processing <filename>
The file has not changed
```
messages (by the way, it would be better to show the relative path instead of just the filename) and then I get the aforementioned message without uploading any new file.
Right now, I have like 90% of the files uploaded but there's some files yet (79 items, totalling 93,7 MB) in one particular folder of a project that refuse to be uploaded.

**Comment by abraunegg** [comment]
With #2 - it will be all about the number of 'versions' of a file that is being stored on OneDrive - this will consume additional space that it shouldn't be.

#2 changes the upload process for Business accounts - can you try that branch rather than master and report back ?

**Comment by jose-sanchezm** [comment]
I've tried with branch Issue-#2 with the same results.

**Comment by abraunegg** [comment]
Can you post the `--verbose` output plus crash details when using #2 branch?

**Comment by jose-sanchezm** [comment]
It's exactly the same as before:

```
Uploading new items of .
std.json.JSONException@std/json.d(1392): Unexpected character '<'. (Line 1:1)
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN""http://www.w3.org/TR/html4/strict.dtd">
<HTML><HEAD><TITLE>Bad Request</TITLE>
<META HTTP-EQUIV="Content-Type" Content="text/html; charset=us-ascii"></HEAD>
<BODY><h2>Bad Request - Invalid URL</h2>
<hr><p>HTTP Error 400. The request URL is invalid.</p>
</BODY></HTML>

----------------
??:? pure @safe void std.json.parseJSON!(char[]).parseJSON(char[], int, std.json.JSONOptions).error(immutable(char)[]) [0x4f8434dc]
??:? pure @safe void std.json.parseJSON!(char[]).parseJSON(char[], int, std.json.JSONOptions).parseValue(ref std.json.JSONValue) [0x4f8440b7]
??:? pure @safe std.json.JSONValue std.json.parseJSON!(char[]).parseJSON(char[], int, std.json.JSONOptions) [0x4f843479]
??:? std.json.JSONValue onedrive.OneDriveApi.perform() [0x4f842d7a]
??:? std.json.JSONValue onedrive.OneDriveApi.get(const(char)[], bool) [0x4f8425fe]
??:? std.json.JSONValue onedrive.OneDriveApi.getPathDetails(const(immutable(char)[])) [0x4f841b43]
??:? void sync.SyncEngine.uploadNewFile(immutable(char)[]) [0x4f84f387]
??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0x4f84eabb]
??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0x4f84e9b2]
??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0x4f84e9b2]
??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0x4f84e9b2]
??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0x4f84e9b2]
??:? void sync.SyncEngine.scanForDifferences(immutable(char)[]) [0x4f84db81]
??:? void main.performSync(sync.SyncEngine, immutable(char)[], bool, bool, bool) [0x4f8345a7]
??:? _Dmain [0x4f833f47]

```
Checking the version number, which is the same as the last commit in #2 branch (I think)

```
jose@jose-pc:~/Programs/onedrive$ onedrive --version
onedrive v1.1.2-14-g491507a

```

**Comment by abraunegg** [comment]
>  it would be better to show the relative path instead of just the filename

Whilst I totally agree, when the application is processing, the 'name' part if received back from OneDrive as part of the JSON received, without any path details.

I could 'add in code' that queries the database to construct the full local logical path for the file, however this certainly would be a performance impediment to the application

**Comment by jose-sanchezm** [comment]
It's ok then. I think it's a good trade-off in performance. I was just trying to figure out what's the file that was causing the failure (and preventing the program from continuing the uploads) but the ones with name are the ones which are correctly uploaded. I don't have a clue (based on the log) of what file is the problematic one. Is there any way to make it even more verbose?

**Comment by jose-sanchezm** [comment]
Thanks. Now at least I know the offending file. This is the output:

```
Uploading new items of .
* Hostname graph.microsoft.com was found in DNS cache
*   Trying 20.190.129.98...
* TCP_NODELAY set
* Connected to graph.microsoft.com (20.190.129.98) port 443 (#3)
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: none
* SSL re-using session ID
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server did not agree to a protocol
* old SSL session ID is stale, removing
* Server certificate:
*  subject: C=US; ST=WA; L=Redmond; O=Microsoft Corporation; OU=Microsoft Corporation; CN=graph.microsoft.com
*  start date: Jan  3 17:32:18 2018 GMT
*  expire date: Jan  3 17:32:18 2020 GMT
*  subjectAltName: host "graph.microsoft.com" matched cert's "graph.microsoft.com"
*  issuer: C=US; ST=Washington; L=Redmond; O=Microsoft Corporation; OU=Microsoft IT; CN=Microsoft IT TLS CA 4
*  SSL certificate verify ok.
> GET /v1.0/me/drive/root:/.%2FProjects%2FDitas%2Fsota%2FState-of-the-art%2C%20challenges%2C%20and%20open%20issues%20in%20the%20integration%20of%20Internet%20of%0AThings%20and%20Cloud%20Computing.pdf:/?select=id,name,eTag,cTag,deleted,file,folder,root,fileSystemInfo,remoteItem,parentReference HTTP/1.1
Host: graph.microsoft.com
User-Agent: Phobos-std.net.curl/2.080 (libcurl/7.60.0)
Accept: */*
Authorization: bearer eyJ0eXAiOiJKV1QiLCJub25jZSI6IkFRQUJBQUFBQUFEWDhHQ2k2SnM2U0s4MlRzRDJQYjdyd1dZNXN3TTJha0IzYW5mX1BkUWl3UV93LXZHN1NaLXdlNHI0Q0FSVGxMQ2VxVW1jRXhjNlRzYzJsX2lWX1pBQWxqWEJqOHFQalBqNTh2d3Y5S05pMENBQSIsImFsZyI6IlJTMjU2IiwieDV0IjoiVGlvR3l3d2xodmRGYlhaODEzV3BQYXk5QWxVIiwia2lkIjoiVGlvR3l3d2xodmRGYlhaODEzV3BQYXk5QWxVIn0.eyJhdWQiOiJodHRwczovL2dyYXBoLm1pY3Jvc29mdC5jb20iLCJpc3MiOiJodHRwczovL3N0cy53aW5kb3dzLm5ldC8zMzQ0MGZjNi1iN2M3LTQxMmMtYmI3My0wZTcwYjAxOThkNWEvIiwiaWF0IjoxNTMwMTA4MjI4LCJuYmYiOjE1MzAxMDgyMjgsImV4cCI6MTUzMDExMjEyOCwiYWNyIjoiMSIsImFpbyI6IlkyZGdZT0R0eU4wOVVUb250SFNwbmNCRkErSC9oNHpEUWkxMXVxY1duL1ArRzlCN1ppTUEiLCJhbXIiOlsicnNhIl0sImFwcF9kaXNwbGF5bmFtZSI6Ik9uZURyaXZlIEZyZWUgQ2xpZW50IiwiYXBwaWQiOiIyMmM0OWEwZC1kMjFjLTQ3OTItYWVkMS04ZjE2M2M5ODI1NDYiLCJhcHBpZGFjciI6IjAiLCJmYW1pbHlfbmFtZSI6IlNhbmNoZXoiLCJnaXZlbl9uYW1lIjoiSm9zZSBBbnRvbmlvIiwiaXBhZGRyIjoiNjIuMTQuMjI4LjE5NiIsIm5hbWUiOiJTYW5jaGV6LCBKb3NlIEFudG9uaW8iLCJvaWQiOiI4MWNjYTdiYi0xYjAyLTQzNGYtYjAxNS04MGM0OTk4ODQxMTYiLCJvbnByZW1fc2lkIjoiUy0xLTUtMjEtMTAxNDE3NjI2MC05ODkzMDcwNy00MDQzNDQ3NzMwLTU5NTM1OSIsInBsYXRmIjoiMTQiLCJwdWlkIjoiMTAwMzdGRkU5QjlDNzBFMCIsInNjcCI6IkZpbGVzLlJlYWRXcml0ZSBGaWxlcy5SZWFkV3JpdGUuQWxsIiwic3ViIjoibjJrRnBBeVladzFkMzhxa2Zvc19zZm1QWE1pV3gycUh5V2dtYTJGQ2RxZyIsInRpZCI6IjMzNDQwZmM2LWI3YzctNDEyYy1iYjczLTBlNzBiMDE5OGQ1YSIsInVuaXF1ZV9uYW1lIjoiam9zZS5zYW5jaGV6bUBhdG9zLm5ldCIsInVwbiI6Impvc2Uuc2FuY2hlem1AYXRvcy5uZXQiLCJ1dGkiOiJWNXVBNmFyN29rMkFhZV9NWE04RUFBIiwidmVyIjoiMS4wIn0.Kz8jgQMI12QWibnzWPxR4sMMUFsocLuv6Q6L8x9LjedwE8W9ZJ78sgHKO6yck82mpmB0Hl3hQBeH0cS6f3G2MxQx8bUoT3thwaKY6Fu-XeMqaYruKYUe4t8DC1IaFITYasKFvj7QD4ImWV9fzrhXAIeWhGe6fErppRhUx-Iw7jWObOvuUvedRFGnJI6uU5eom82UTM30rsglo8GKyCyazHU4_lP-AGiZfca1aLTg-drZI4-FirvcLJVCTX4sFwVDSThv_eH7WKEOYRDDjgOs-S32zl5gyXi7ru4dJyqm0VckCYO9QD42sPVlRWHEVrMYcIMOZvNiwXqB_FMo2Bp4Rg

< HTTP/1.1 400 Bad Request
< Content-Type: text/html; charset=us-ascii
< Server: Microsoft-HTTPAPI/2.0
< Date: Wed, 27 Jun 2018 14:08:52 GMT
< Connection: close
< Content-Length: 324
< 
* Closing connection 3
std.json.JSONException@std/json.d(1392): Unexpected character '<'. (Line 1:1)
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN""http://www.w3.org/TR/html4/strict.dtd">
<HTML><HEAD><TITLE>Bad Request</TITLE>
<META HTTP-EQUIV="Content-Type" Content="text/html; charset=us-ascii"></HEAD>
<BODY><h2>Bad Request - Invalid URL</h2>
<hr><p>HTTP Error 400. The request URL is invalid.</p>
</BODY></HTML>

----------------
??:? pure @safe void std.json.parseJSON!(char[]).parseJSON(char[], int, std.json.JSONOptions).error(immutable(char)[]) [0xaa13d4dc]
??:? pure @safe void std.json.parseJSON!(char[]).parseJSON(char[], int, std.json.JSONOptions).parseValue(ref std.json.JSONValue) [0xaa13e0b7]
??:? pure @safe std.json.JSONValue std.json.parseJSON!(char[]).parseJSON(char[], int, std.json.JSONOptions) [0xaa13d479]
??:? std.json.JSONValue onedrive.OneDriveApi.perform() [0xaa13cd7a]
??:? std.json.JSONValue onedrive.OneDriveApi.get(const(char)[], bool) [0xaa13c5fe]
??:? std.json.JSONValue onedrive.OneDriveApi.getPathDetails(const(immutable(char)[])) [0xaa13bb43]
??:? void sync.SyncEngine.uploadNewFile(immutable(char)[]) [0xaa149387]
??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0xaa148abb]
??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0xaa1489b2]
??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0xaa1489b2]
??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0xaa1489b2]
??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0xaa1489b2]
??:? void sync.SyncEngine.scanForDifferences(immutable(char)[]) [0xaa147b81]
??:? void main.performSync(sync.SyncEngine, immutable(char)[], bool, bool, bool) [0xaa12e5a7]
??:? _Dmain [0xaa12df47]

```
Is there something wrong with this file name?

**Comment by abraunegg** [comment]
So nothing appears wrong with the file name, there are no invalid characters there, however this is telling:

`old SSL session ID is stale, removing`

Potentially something around that which is causing the request to fail. This gives me some direction to investigate.

**Comment by abraunegg** [comment]
I am having real difficulty here in reproducing this.

I have uploaded ~2000 files to OneDrive Business, running through a script to modify them so that I am simulating 'user activity'

The only difference I see here is the version of Curl - I am running `libcurl/7.29.0` vs your platform `libcurl/7.60.0`

What OS & version are you running? Can you also provide the output of `curl -V` ? What OpenSSL version are you using?

My versions:
```
curl 7.29.0 (x86_64-redhat-linux-gnu) libcurl/7.29.0 NSS/3.19.1 Basic ECC zlib/1.2.3 libidn/1.18 libssh2/1.4.2
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp scp sftp smtp smtps telnet tftp 
Features: AsynchDNS GSS-Negotiate IDN IPv6 Largefile NTLM NTLM_WB SSL libz 
```
```
OpenSSL 1.0.1e-fips 11 Feb 2013
```

**Comment by jose-sanchezm** [comment]
I'm using arch linux with curl 7.60.0 and OpenSSL 1.1.0.h

**Comment by abraunegg** [comment]
OK .. let's step through this

1. Getting a 504 Gateway Timeout. What is your network topology like? Do you have any gateway filtering / transparent proxy going on? To get - there has to be some sort of issue getting to the DC where your attempting to connect to. It would have been good to get a full verbose log for the error including where the 504 was received as I thought I had resolved most of those to gracefully fail & retry.

2. Failing on uploading in the same place .. What I am going to do is build an Arch Linux system and see if I can somehow replicate this with the path names & so on.

**Comment by abraunegg** [comment]
I have installed Arch Linux and attempted to replicate the issue here locally:
```
uname -a
Linux 127.0.0.1localhost 4.17.2-1-ARCH #1 SMP PREEMPT Sat Jun 16 11:08:59 UTC 2018 x86_64 GNU/Linux

curl -V
curl 7.60.0 (x86_64-pc-linux-gnu) libcurl/7.60.0 OpenSSL/1.1.0h zlib/1.2.11 libidn2/2.0.5 libpsl/0.20.2 (+libidn2/2.0.4) nghttp2/1.31.1
Release-Date: 2018-05-16
Protocols: dict file ftp ftps gopher http https imap imaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets HTTPS-proxy PSL

openssl version
OpenSSL 1.1.0h  27 Mar 2018
```

Currently testing with a set of files ~450Mb containing ~3K random files including images (jpg, png, psd), code files (c, c#, d, php), shell scrips, PDF's, business documents (PPT, DOC, XLS etc), plain text files, archives (rar, zip, 7zip, tar.gz etc).

**Comment by abraunegg** [comment]
**Testing Results:**
Master branch (onedrive v1.1.2-19-g63c5dd8) - 100% successful upload
Issue #2 branch (onedrive v1.1.2-15-g9ec5885) - 100% successful upload 

Based on this testing, I suspect there can only be 2 things going on here:

1. Something environmental going on with your system / access to OneDrive itself that is scanning or doing something with the outbound HTTPS communication.
2. The actual file is the issue. If the file is not sensitive, please can you archive it in a password protected archive & email to support@mynas.com.au. If I can replicate with the same file (after double successful tests above), then the issue lies with the file & how OneDrive is handling the file with its internal mechanics.

**Comment by jose-sanchezm** [comment]
I've sent you the file. It's not sensitive as it can be found easily on the internet.

**Comment by abraunegg** [comment]
Thanks for the file. I've tested the upload using the same path & file:

```
Loading config ...
Using Config Dir: /root/.config/onedrive
No config file found, using defaults
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /root/OneDrive
Initializing the Synchronization Engine ...
Account Type: business
Default Drive ID: <redacted>
Default Root ID: <redacted>
Remaining Free Space: 1097162618794
Fetching details for OneDrive Root
OneDrive Root exists in the database
Applying changes of Path ID: <redacted>
Uploading differences of .
Processing root
The directory has not changed
Uploading new items of .
OneDrive Client requested to create remote path: ./Projects
The requested directory to create was not found on OneDrive - creating remote directory: ./Projects
Successfully created the remote directory ./Projects on OneDrive
OneDrive Client requested to create remote path: ./Projects/Ditas
The requested directory to create was not found on OneDrive - creating remote directory: ./Projects/Ditas
Successfully created the remote directory ./Projects/Ditas on OneDrive
OneDrive Client requested to create remote path: ./Projects/Ditas/sota
The requested directory to create was not found on OneDrive - creating remote directory: ./Projects/Ditas/sota
Successfully created the remote directory ./Projects/Ditas/sota on OneDrive
Uploading file ./Projects/Ditas/sota/State-of-the-art, challenges, and open issues in the integration of Internet ofThings and Cloud Computing.pdf ...
Uploading fragment: 0-508719/508719
 done.
Remaining free space: 1097162110075
Applying changes of Path ID: <redacted>
```

The MS DC I was hitting was:

```
x-ms-ags-diagnostic: {"ServerInfo":{"DataCenter":"Australia Southeast","Slice":"SliceC","Ring":"2","ScaleUnit":"000","Host":"AGSFE_IN_0","ADSiteName":"ASE"}}
```

So this rules out the file itself - so it has to be something environmental to your machine, your access path to MS OneDrive or something along those lines.

Things potentially to look at:
 * Network MTU
* Packet fragmentation
* Packet loss

**Comment by jose-sanchezm** [comment]
There was a weird character that was not visible in the name of the file (but it appeared when trying to rename it in Nautilus) that probably was removed in the file I sent you. Removing this character allowed it to continue.

**Comment by abraunegg** [comment]
Is there anyway to get / find the file in your archives that has the weird character? It would be great to have that to use & test against.

**Comment by jose-sanchezm** [comment]
By chance I have a copy pre-renaming. Looking at how the title is pasted here, it seems that there is a return character between "of" and "Things". Compressing it to a tar.xz archive seems to conserve the name so I'm sending it to you by email.

**Comment by abraunegg** [comment]
Thanks - this will give me something to test against & provide a fix for

**Comment by abraunegg** [comment]
Can you help test PR #41 ?

**Testing results:**

```
Loading config ...
Using Config Dir: /root/.config/onedrive
No config file found, using defaults
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /root/OneDrive
Initializing the Synchronization Engine ...
Account Type: business
Default Drive ID: <redacted>
Default Root ID: <redacted>
Remaining Free Space: 1097161599398
Fetching details for OneDrive Root
OneDrive Root exists in the database
Applying changes of Path ID: <redacted>
Uploading differences of .
Processing root
The directory has not changed
Processing State-of-the-art, challenges, and open issues in the integration of Internet ofThings and Cloud Computing-good.pdf
The file has not changed
Uploading new items of .
Skipping item - invalid name (Contains an invalid whitespace item): ./State-of-the-art, challenges, and open issues in the integration of Internet of
Things and Cloud Computing.pdf
Applying changes of Path ID: <redacted>
```

**Comment by abraunegg** [comment]
Merged PR #41
